### DISCUSSION TITLE
[Support] 怎么将 midwayjs 标准应用部署到 aws lambda?

### DISCUSSION THREAD

**Discussion by zuozuomuxi** [midwayjs/midway#3541] [Q&A] [answered]
### Describe the problem（描述问题）

我想将我的 midwayjs 标准应用部署到 aws lambda

如果是直接使用 koa 或者 express 框架的应用，可以用 [CodeGenieApp/serverless-express](https://github.com/CodeGenieApp/serverless-express) 这个项目包装一下

```js
const serverlessExpress = require('@codegenie/serverless-express')
const app = require('./app')
exports.handler = serverlessExpress({ app })
```

如果是 nestjs 这种框架，也可以通过类似的方法做，因为 nestjs 接受从外部传入 koa 或者 express 的实例，可以看这个[例子](https://github.com/CodeGenieApp/serverless-express/blob/mainline/examples/basic-starter-nestjs/src/lambda.ts)

但是 midwayjs 的 express 和 koa 适配模块不支持从外部传入 express 和 koa 的实例，就没法做这个事情了。

像 `@midwayjs/koa` 就是直接在内部 new koa(...)
https://github.com/midwayjs/midway/blob/addbfe0cecac0347ed005c3184249c5d4f45ebad/packages/web-koa/src/framework.ts#L74

那么有什么其他方法能让我将标准应用部署到 aws lambda 呢？或者能否改造下 `@midwayjs/koa` 和 `@midwayjs/express`，让它能接受外部传入的实例

### Midway Versions（Midway 版本）

3.13.8

**Comment by czy88840616** [comment]
我理解只要能拿到 koa/express 的 app 实例，传递到 serverlessExpress 就行？

**Reply by czy88840616** [reply, marked as answer]
初始化完从 `framework.getApplication()` 拿到 app 就行。

```js
// bootstrap.js
const { Bootstrap } = require('@midwayjs/bootstrap');
const { Framework } = require('@midwayjs/koa');
const serverlessExpress = require('@codegenie/serverless-express')

let serverlessExpressInstance;

async function setup (event, context) {

  // 初始化 miwday，拿到全局容器
  const container = await Bootstrap.run();
  // 拿到框架实例
  const framework = container.get(Framework);
  // 拿到 app
  const app = framework.getApplication();

  // 下面都是 serverlessExpress 文档上的异步获取用法
  serverlessExpressInstance = serverlessExpress({ app })
  return serverlessExpressInstance(event, context)
}

function handler (event, context) {
  if (serverlessExpressInstance) return serverlessExpressInstance(event, context)
  return setup(event, context)
}

module.exports = handler;
```

**Reply by zuozuomuxi** [reply]
我不确定行不行，你代码里 setup 是异步的，但是 handler 可能需要同步返回。我待会尝试一下。
是否有方法能同步的拿到 app 呢？

**Reply by czy88840616** [reply]
![image](https://github.com/midwayjs/midway/assets/418820/59981abf-6fd0-4fe2-a1f5-b3e3786188be)
这是serverlessExpress 官方 readme 自己的示例。。

**Reply by zuozuomuxi** [reply]
@czy88840616 刚没看到，我去部署尝试了一下，确实是可以的，已经跑起来了，感谢。

部署时遇到了另一个小问题，顺便提一下。lambda 运行环境中没有 HOME 这个环境变量，导致 appInfo.root 的值是 undefined，`@midwayjs/logger` 会用到 path.join(appInfo.root, ...) 来计算写入日志文件的路径， path.join 不接收 undefined 值，这里就会出错。

这个问题通过配置日志路径就可以解决，只是要注意下只能配置为 /tmp/ 开头的路径，lambda 只有 tmp 是可写入的

**Reply by czy88840616** [reply]
serverless环境都是如此，包括阿里云，腾讯云，日志目录都要调整或者关闭文本写入
